daffodil-test/src/test/resources/org/apache/daffodil/layers/IPv4.dfdl.xsd

```diff
@@ -0,0 +1,199 @@
+        <xs:element name="FragmentOffset" type="tns:bit" dfdl:length="13"/>
+        <xs:element name="TTL" type="tns:bit" dfdl:length="8"/>
+        <xs:element name="Protocol" type="tns:bit" dfdl:length="8"/>
+        <xs:element name="Checksum" type="chksum:IPv4Checksum"/>
+        <xs:element name="IPSrc" type="tns:hexByte" dfdl:length="4"/>
+        <xs:element name="IPDest" type="tns:hexByte" dfdl:length="4"/>
+      </xs:sequence>
+      </xs:sequence>
+      <!--
+      This exists purely to make unparsing of the header suspend when trying to unparse
+      the Length element.
+
+      We just want the tests to rule out deadlock in this case.
+      -->
+      <xs:element name="RealLength" type="tns:bit"
+                  dfdl:inputValueCalc='{ ../Length }'/>
+      <!--
+      We want the schema author to have all options on what to do if the
+      checksum is incorrect when parsing. Hence, we just put the computed value
+      into this element.
+      -->
+      <xs:element name="ComputedChecksum" type="chksum:IPv4Checksum"
+                  dfdl:inputValueCalc='{ $chksum:IPv4Checksum }'/>
+      <!--
+      One recommendation is to treat incorrect checksum values as a validation
+      error. This preserves the ability to use the schema forensically to examine
+      data with incorrect checksums.
+
+      This uses a DFDL recoverable error assertion to report that the
+      checksum is incorrect. A recoverable error assert is not technically
+      a validation error, but behaves similarly in that it does not prevent the
+      parse from completing. It is essentially a warning about the data.
+
+      A schematron check would accomplish much the same thing, with the
+      advantage that the error would be reported as an "official" validation error.
+      -->
+      <xs:sequence>
+        <xs:annotation>
+          <xs:appinfo source="http://www.ogf.org/dfdl/">
+            <dfdl:assert
+              failureType="recoverableError"
+              message="Incorrect checksum."
+              test='{ Checksum eq ComputedChecksum }'
+            />
+          </xs:appinfo>
+        </xs:annotation>
+      </xs:sequence>
+    </xs:sequence>
+  </xs:complexType>
+
+  <xs:simpleType name="bit"
+                 dfdl:lengthKind="explicit"
+                 dfdl:lengthUnits="bits">
+    <xs:restriction base="xs:unsignedInt"/>
+  </xs:simpleType>
+
+  <xs:simpleType name="hexByte" dfdl:lengthKind="explicit" dfdl:lengthUnits="bytes">
+    <xs:restriction base="xs:hexBinary"/>
+  </xs:simpleType>
+
+  <xs:simpleType name="uint16" dfdl:lengthKind="explicit" dfdl:length="16">
+    <xs:restriction base="xs:unsignedInt"/>
+  </xs:simpleType>
+
+  <xs:simpleType name="uint32" dfdl:lengthKind="explicit" dfdl:length="32">
+    <xs:restriction base="xs:unsignedInt"/>
+  </xs:simpleType>
+
+  <xs:simpleType name="int32" dfdl:lengthKind="explicit" dfdl:length="4" dfdl:lengthUnits="bytes">
```

Comment:
I would expect to read 32 bits, not 4 bytes, in an `int32` definition, but you've made this definition (almost) explicit enough to be portable to any schema.  Actually, the funny thing is that the base type `xs:int` is also 32 bits long so the only restrictions you really need are `dfdl:lengthKind="explicit"` and `dfdl:representation="binary"`.